cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregatorMetadata.java

```diff
@@ -0,0 +1,255 @@
+ */
+
+package com.linkedin.kafka.cruisecontrol.monitor.sampling.aggregator;
+
+import com.linkedin.cruisecontrol.common.LongGenerationed;
+import com.linkedin.cruisecontrol.monitor.sampling.aggregator.SnapshotAndImputation;
+import com.linkedin.cruisecontrol.monitor.sampling.aggregator.WindowAggregationResult;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentSkipListMap;
+import org.apache.kafka.common.TopicPartition;
+
+
+/**
+ * The metadata for {@link KafkaMetricSampleAggregator}. It helps maintain the completeness information of each
+ * snapshot window.
+ */
+class KafkaMetricSampleAggregatorMetadata extends LongGenerationed {
+  private final Set<String> _invalidTopics;
+  private final Set<String> _topicsWithTooManyImputations;
+  private final Map<Long, WindowCompleteness> _completenessByWindows;
+  private final double _maxAllowedImputations;
+
+  private volatile int _clusterGeneration = -1;
+  private volatile int _numValidPartitions = 0;
+
+  KafkaMetricSampleAggregatorMetadata(long generation, double maxAllowedImputations) {
+    super(generation);
+    _topicsWithTooManyImputations = new HashSet<>();
+    _invalidTopics = new HashSet<>();
+    _completenessByWindows = new ConcurrentSkipListMap<>(Comparator.reverseOrder());
+    _maxAllowedImputations = maxAllowedImputations;
+  }
+
+  boolean generationsMatch(long generation, int clusterGeneration) {
+    return compareGeneration(generation) == 0 && _clusterGeneration == clusterGeneration;
+  }
+
+  /**
+   * Update the window completeness.
+   */
+  synchronized void update(long generation,
+                           int clusterGeneration,
+                           List<WindowAggregationResult<TopicPartition>> windowResults) {
+    if (!generationsMatch(generation, clusterGeneration)) {
+      updateInvalidTopics(windowResults);
+      updateImputations(windowResults);
+      long maxGeneration = updateWindowCompleteness(windowResults);
+      setGeneration(maxGeneration);
+      _clusterGeneration = clusterGeneration;
+    }
+  }
+
+  synchronized int numValidPartitions() {
+    return _numValidPartitions;
+  }
+
+  /**
+   * Get the available windows and valid snapshots that meets the monitored percentage requirements.
+   * @param minMonitoredPartitionsPercentage the minimum required monitored percentage.
+   * @param totalNumPartitions the total number of partitions.
+   * @return An {@link AvailableWindows} containing the information of windows that meets the monitored 
+   * partitions percentage requirements.
+   */
+  synchronized AvailableWindows validWindows(double minMonitoredPartitionsPercentage, 
```

Comment:
Shouldn't this logic be extracted from here and put in core?  The problem we have with consistent metrics collection is likely not limited to Kafka.

Reply:
Kafka have a special logic that if all the partitions should be included/excluded all together. The generic logic assumes independence of the entities. So we need special logic here.